examples/fabric-admin/device_manager/DeviceSubscriptionManager.h

```diff
@@ -0,0 +1,88 @@
+/*
+ *   Copyright (c) 2024 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+#pragma once
+
+#include <app/ReadClient.h>
+#include <controller/CHIPDeviceController.h>
+#include <lib/core/DataModelTypes.h>
+
+#include <memory>
+
+#include "fabric_bridge_service/fabric_bridge_service.pb.h"
+#include "fabric_bridge_service/fabric_bridge_service.rpc.pb.h"
+
+class DeviceSubscriptionManager
+{
+public:
+    /// Attribute subscription to attributes that are important to keep track and send to fabric-bridge
+    /// via RPC when change has been identified.
+    ///
+    /// An instance of DeviceSubscription is intended to be used only once. Once a DeviceSubscription is
+    /// terminal, either from an error or from subscriptions getting shut down, we expect the instance
+    /// to be deleted. Any new subscription should instantiate another instance of DeviceSubscription.
+    ///
+    /// class is public because OnDeviceConnectedWrapper and OnDeviceConnectionFailureWrapper need
+    /// cast back to DeviceSubscription.
+    class DeviceSubscription : public chip::app::ReadClient::Callback
```

Comment:
The idea is that nothing but only the DeviceSubscriptionManager should instantiate an instance of DeviceSubscription. This allows us to give DeviceSubscription access to private member methods of DeviceSubscriptionManager which help enforce via code that DeviceSubscriptionTerminated is only ever called by DeviceSubscription without the use of `friend`, which and others argue the use of `friend` is an anti-pattern

Reply:
The DeviceSubscription class is tightly coupled with DeviceSubscriptionManager. This could make it difficult to use DeviceSubscription independently or to mock it for testing.

We have a lot of similar design throughout the SDK, I see use nested class is very rare, and only for extremely simple case. The my other suggestions on decoupling